cmd/autoexport.go

```diff
@@ -27,26 +36,109 @@ func AutoExport(c *cli.Context) {
 	defer cli.Close()
 	cli.NegotiateAPIVersion(ctx)
 
-	b := backend.NewBackend(cli)
+	b := docker.NewDockerBackend(cli, promNetwork, models.NewPredefinedExporterFinder())
 	cancellables := newCancellableCollection()
+	logger := log.GetLogger(ctx)
 
-	logrus.Info("Removing stale exporters...")
+	logger.Info("Removing stale exporters...")
+	if err = b.CleanupExporters(ctx, forceRecreate); err != nil {
+		logger.Error("Cleanup failed: ", err)
+	}
 
-	if err := b.CleanupAllExporters(ctx, forceRecreate); err != nil {
+	logger.Info("Start listening for new backend events...")
+	taskEvtCh := make(chan models.TaskEvent)
+	go b.ListenForTasksToExport(ctx, taskEvtCh)
+
+	logger.Info("Starting missing exporters...")
+	missingExporters, err := b.FindMissingExporters(ctx)
+	if err != nil {
 		logrus.Errorf("%+v", err)
 	}
 
-	logrus.Info("Starting missing exporters...")
+	for _, exporter := range missingExporters {
+		cancellableID := fmt.Sprintf("%s.%s", exporter.ExportedTask.Name, exporter.ExporterType)
+		ctx := cancellables.add(cancellableID, ctx)
+		handler := startHandlerFactory(b, exporter)
 
-	if err := b.StartMissingExporters(ctx, promNetwork); err != nil {
-		logrus.Errorf("%+v", err)
+		logger := logger.WithFields(logrus.Fields{
+			"operation":   "start",
+			"exporter":    log.FormatExporterField(exporter),
+			"task":        log.FormatTaskField(exporter.ExportedTask),
+			"retry_count": 0,
+		})
+		ctx = log.WithLogger(ctx, logger)
+
+		go retry(cancellables, ctx, 3, 30 * time.Second, handler)
```

Comment:
do you have to pass the whole `cancellable` collection to the `retry` method ? isn't this method call only used for the current missing exporter handling ?

Reply:
Both the startup and cleanup operations are retried if they fail. Both can be cancelled if another operation has to run for the same exporter. Thus, the cancellable collection is passed to the retry method so the canceler is removed from the collection. So yeah, it's needed in both cases.